src/source.rs

```diff
@@ -103,65 +115,45 @@ impl FileSource {
         )
     }
 
-    pub fn fetch_file_sync(&self, langid: &LanguageIdentifier, path: &str) -> ResourceOption {
+    pub fn fetch_file_sync(&self, langid: &LanguageIdentifier, path: &str) -> RcResourceOption {
+        use ResourceStatus::*;
+
         let full_path = self.get_path(langid, &path);
 
-        let mut cache = self.cache.try_borrow_mut().unwrap();
-        let res = cache.entry(full_path.clone()).or_insert_with(|| {
-            (self.fetch_sync)(&full_path)
-                .expect("I/O Error")
+        // println!("[l10nregistry] fetch_file_sync: {:?}", full_path);
+
+        let res = self.shared.lookup_resource(full_path.clone(), || {
+            self.shared
+                .fetcher
+                .fetch_sync(&full_path)
+                .ok()
                 .map(|source| Rc::new(FluentResource::try_new(source).unwrap()))
                 .into()
         });
 
         match res {
-            ResourceStatus::Value(res) => Some(res.clone()),
-            ResourceStatus::Async(..) => {
-                println!("[l10nregistry] Attempting to synchronously load file {} while it's being loaded asynchronously.", &full_path);
-                cache.remove(&full_path);
-                drop(cache);
-                self.fetch_file_sync(langid, path)
+            Missing => None,
+            Loaded(res) => Some(res),
+            Loading(..) => {
+                panic!("[l10nregistry] Attempting to synchronously load file {} while it's being loaded asynchronously.", &full_path);
```

Comment:
could we here `debug_print` and drop the async load, and replace it with a sync load?

Reply:
Or we could perform a sync load and not update the cache? I wasn't sure which was the best approach.

I wanted to be able to replace the loading state with loaded and fire the value to the waiting futures, but I'm not sure how to do that. A potential solution would be to store a `oneshot::Sender` and `Shared<oneshot::Receiver>`. The async loading future could then check that the state is still loading, and if so, `send` the result. The sync loading could do the same.

It's a little more complicated and I just wanted to get the system working. Could be leave this and open a issue to implement a more desirable behavior?